File: frameworks/rust/poem/Implementation/routes/items.rs
use poem::http::header::LOCATION;
use poem::http::{StatusCode, Uri};
use poem::web::{Data, Json, Path};
use poem::{EndpointExt, IntoResponse, Response, Route, get, handler, post};
use serde::Deserialize;

use crate::payloads::Item;
use crate::{Payloads, Shared};

/// An item as a client creates or replaces one.
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct NewItem {
    name: String,
    category: String,
    price_cents: i64,
    in_stock: bool,
}

impl NewItem {
    fn at(self, id: i64) -> Item {
        Item { id, name: self.name, category: self.category, price_cents: self.price_cents, in_stock: self.in_stock }
    }
}

/// The two fields items.update changes.
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ItemPatch {
    price_cents: Option<i64>,
    in_stock: Option<bool>,
}

impl ItemPatch {
    fn onto(self, row: &Item) -> Item {
        Item { price_cents: self.price_cents.unwrap_or(row.price_cents), in_stock: self.in_stock.unwrap_or(row.in_stock), ..row.clone() }
    }
}

// rb:handler items.read,items.head,errors.not_found
/// The row with the id in the path. poem answers HEAD with this handler and leaves the body
/// unwritten.
#[handler]
fn read(Path(id): Path<i64>, Data(&p): Data<&Shared>) -> Result<Json<&'static Item>, StatusCode> {
    p.row(id).map(Json).ok_or(StatusCode::NOT_FOUND)
}

// rb:handler items.create
/// 201, where the item would live, and the item under the id after the last row. The Location is
/// built from the request's path.
#[handler]
fn create(uri: &Uri, Json(item): Json<NewItem>, Data(&p): Data<&Shared>) -> Response {
    let created = item.at(p.large.count + 1);
    let location = format!("{}/{}", uri.path(), created.id);
    Json(created).with_status(StatusCode::CREATED).with_header(LOCATION, location).into_response()
}

// rb:handler items.replace
#[handler]
fn replace(Path(id): Path<i64>, Json(item): Json<NewItem>) -> Json<Item> {
    Json(item.at(id))
}

// rb:handler items.update
#[handler]
fn update(Path(id): Path<i64>, Json(change): Json<ItemPatch>, Data(&p): Data<&Shared>) -> Result<Json<Item>, StatusCode> {
    p.row(id).map(|row| Json(change.onto(row))).ok_or(StatusCode::NOT_FOUND)
}

// rb:handler items.delete
#[handler]
fn remove(Path(id): Path<i64>, Data(&p): Data<&Shared>) -> StatusCode {
    if p.row(id).is_some() { StatusCode::NO_CONTENT } else { StatusCode::NOT_FOUND }
}

/// items: every method on one resource over the rows of items.large. A measured row may not leave
/// the server changed, so the writes store nothing and answer as if they had written. poem's method
/// router answers a method the path has no endpoint for with 405.
pub fn add(route: Route, p: &'static Payloads) -> Route {
    route
        .at("/items", post(create.data(p)))
        .at("/items/:id", get(read)
            .put(replace).patch(update).delete(remove).data(p))
}

File: frameworks/go/net-http/Implementation/items.go
package implementation

import (
	"encoding/json"
	"net/http"
	"strconv"
)

// NewItem is an item as a client creates or replaces one.
type NewItem struct {
	Name       string `json:"name"`
	Category   string `json:"category"`
	PriceCents int    `json:"priceCents"`
	InStock    bool   `json:"inStock"`
}

func (n NewItem) at(id int) Item {
	return Item{ID: id, Name: n.Name, Category: n.Category, PriceCents: n.PriceCents, InStock: n.InStock}
}

// ItemPatch is the two fields items.update changes. A field the body leaves out stays nil.
type ItemPatch struct {
	PriceCents *int  `json:"priceCents"`
	InStock    *bool `json:"inStock"`
}

func (patch ItemPatch) onto(row Item) Item {
	if patch.PriceCents != nil {
		row.PriceCents = *patch.PriceCents
	}
	if patch.InStock != nil {
		row.InStock = *patch.InStock
	}
	return row
}

// itemsRoutes are every method on one resource over the rows of items.large. A measured row may
// not leave the server changed, so the writes store nothing and answer as if they had written.
// The ServeMux answers a method the path has no pattern for with 405 and an Allow header.
func itemsRoutes(mux *http.ServeMux, p *Payloads) {
	// A GET pattern matches HEAD too, and net/http leaves the body unwritten.
	// rb:handler items.head
	mux.HandleFunc("GET /items/{id}", func(w http.ResponseWriter, r *http.Request) {
		id, err := strconv.Atoi(r.PathValue("id"))
		if err != nil {
			refuse(w, err)
			return
		}
		row, found := p.Row(id)
		if !found {
			http.NotFound(w, r)
			return
		}
		respond(w, http.StatusOK, row)
	})

	// Payload's json:"items" tag also reads as the route literal, so the route is marked.
	// rb:handler items.create
	mux.HandleFunc("POST /items", func(w http.ResponseWriter, r *http.Request) {
		var item NewItem
		if err := json.NewDecoder(r.Body).Decode(&item); err != nil {
			refuse(w, err)
			return
		}
		created := item.at(p.Large.Count + 1)
		w.Header().Set("Location", "/items/"+strconv.Itoa(created.ID))
		respond(w, http.StatusCreated, created)
	})

	mux.HandleFunc("PUT /items/{id}", func(w http.ResponseWriter, r *http.Request) {
		id, err := strconv.Atoi(r.PathValue("id"))
		if err != nil {
			refuse(w, err)
			return
		}
		var item NewItem
		if err := json.NewDecoder(r.Body).Decode(&item); err != nil {
			refuse(w, err)
			return
		}
		respond(w, http.StatusOK, item.at(id))
	})

	mux.HandleFunc("PATCH /items/{id}", func(w http.ResponseWriter, r *http.Request) {
		id, err := strconv.Atoi(r.PathValue("id"))
		if err != nil {
			refuse(w, err)
			return
		}
		row, found := p.Row(id)
		if !found {
			http.NotFound(w, r)
			return
		}
		var patch ItemPatch
		if err := json.NewDecoder(r.Body).Decode(&patch); err != nil {
			refuse(w, err)
			return
		}
		respond(w, http.StatusOK, patch.onto(row))
	})

	mux.HandleFunc("DELETE /items/{id}", func(w http.ResponseWriter, r *http.Request) {
		id, err := strconv.Atoi(r.PathValue("id"))
		if err != nil {
			refuse(w, err)
			return
		}
		if _, found := p.Row(id); !found {
			http.NotFound(w, r)
			return
		}
		w.WriteHeader(http.StatusNoContent)
	})
}
